src/hotspot/share/runtime/crac.cpp

```diff
@@ -699,6 +701,64 @@ Handle crac::checkpoint(jarray fd_arr, jobjectArray obj_arr, bool dry_run, jlong
   return ret_cr(JVM_CHECKPOINT_ERROR, Handle(), Handle(), codes, msgs, THREAD);
 }
 
+bool crac::is_image_score_supported() {
+  return _engine->prepare_image_score_api() == CracEngine::ApiStatus::OK;
+}
+
+bool crac::record_image_score(jobjectArray metrics, jdoubleArray values) {
+  if (_engine->prepare_image_score_api() != CracEngine::ApiStatus::OK) {
+    // silently ignoring this
+    return true;
+  }
+  ResourceMark rm;
+  objArrayOop metrics_oops = objArrayOop(JNIHandles::resolve_non_null(metrics));
+  typeArrayOop values_oops = typeArrayOop(JNIHandles::resolve_non_null(values));
+  assert(metrics_oops->length() == values_oops->length(), "should be equal");
+  for (int i = 0; i < metrics_oops->length(); ++i) {
+    oop metric_oop = metrics_oops->obj_at(i);
+    assert(metric_oop != nullptr, "not null");
+    const char* metric = java_lang_String::as_utf8_string(metric_oop);
+    double value = values_oops->double_at(i);
+    if (!_engine->set_score(metric, value)) {
+      return false;
+    }
+  }
+
+  bool result = true;
+  result = result && _engine->set_score("java.lang.Runtime.availableProcessors", os::active_processor_count());
+  result = result && _engine->set_score("java.lang.Runtime.totalMemory", Universe::heap()->capacity());
+  result = result && _engine->set_score("java.lang.Runtime.maxMemory", Universe::heap()->max_capacity());
+
+  double uptime = TimeHelper::counter_to_millis(os::elapsed_counter());
+  result = result && _engine->set_score("vm.boot_time", os::javaTimeMillis() - uptime);
+  result = result && _engine->set_score("vm.uptime", uptime);
+  result = result && _engine->set_score("vm.uptime_since_restore", TimeHelper::counter_to_millis(os::elapsed_counter_since_restore()));
+
+#if INCLUDE_MANAGEMENT
+  jlong shared_loaded_classes = ClassLoadingService::loaded_shared_class_count();
+  jlong shared_unloaded_classes = ClassLoadingService::unloaded_shared_class_count();
+  // The keys match what jcmd <pid> PerfCounter.print would use
+  result = result && _engine->set_score("java.cls.loadedClasses", ClassLoadingService::loaded_class_count() - shared_loaded_classes);
+  result = result && _engine->set_score("java.cls.sharedLoadedClasses", shared_loaded_classes);
+  result = result && _engine->set_score("java.cls.unloadedClasses", ClassLoadingService::unloaded_class_count() - shared_unloaded_classes);
+  result = result && _engine->set_score("java.cls.sharedUnloadedClasses", shared_unloaded_classes);
+#endif // INCLUDE_MANAGEMENT
+  result = result && _engine->set_score("sun.cls.appClassLoadCount", ClassLoader::perf_app_classload_count()->get_value());
```

Comment:
`ClassLoader::perf_app_classload_count()` should return null when `UsePerfData` is false. Worth to check the other counters too.

Reply:
Good catch, there should be a test variant with `-XX:-UsePerfData`